stl/src/print.cpp

```diff
@@ -0,0 +1,149 @@
+        }
+
+        const HANDLE _Console_handle = reinterpret_cast<HANDLE>(_get_osfhandle(_Fd));
+
+        if (_Console_handle == INVALID_HANDLE_VALUE) [[unlikely]] {
+            return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Invalid_parameter};
+        }
+
+        // We can check if _Console_handle actually refers to a console or not by checking the
+        // return value of GetConsoleMode().
+        DWORD _Console_mode;
+        const bool _Is_unicode_console = (GetConsoleMode(_Console_handle, &_Console_mode) != 0);
+
+        if (!_Is_unicode_console) {
+            return __std_unicode_console_retrieval_result{._Error = __std_win_error::_File_not_found};
+        }
+
+        return __std_unicode_console_retrieval_result{
+            ._Console_handle = static_cast<__std_unicode_console_handle>(
+                reinterpret_cast<_STD underlying_type_t<__std_unicode_console_handle>>(_Console_handle))};
+    }
+} // unnamed namespace
+
+_EXTERN_C
+
+[[nodiscard]] _Success_(return._Error == __std_win_error::_Success) __std_unicode_console_retrieval_result
+    __stdcall __std_get_unicode_console_handle_from_file_stream(_In_ const __std_file_stream_pointer _Stream) noexcept {
+    if (_Stream == __std_file_stream_pointer::_Invalid) [[unlikely]] {
+        return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Invalid_parameter};
+    }
+
+    FILE* const _File_stream_ptr = reinterpret_cast<FILE*>(_Stream);
+    return _Get_console_handle_from_file_stream(_File_stream_ptr);
+}
+
+_END_EXTERN_C
+
+namespace {
+    [[nodiscard]] _STD expected<_STD wstring, __std_win_error> __stdcall _Transcode_utf8_string(
+        const char* const _Str, const size_t _Str_size) noexcept {
+        // MultiByteToWideChar() fails if strLength == 0.
+        if (_Str_size == 0) [[unlikely]] {
+            return {};
+        }
+
+        // The C++ specifications for vprint_unicode() suggest that we replace invalid
+        // code units with U+FFFD. This is done automatically by MultiByteToWideChar(),
+        // so long as we do not use the MB_ERR_INVALID_CHARS flag.
+        const int32_t _Num_chars_required =
+            MultiByteToWideChar(CP_UTF8, 0, _Str, static_cast<int>(_Str_size), nullptr, 0);
+
+        if (_Num_chars_required == 0) [[unlikely]] {
+            return _STD unexpected{static_cast<__std_win_error>(GetLastError())};
+        }
+
+        _STD wstring _Wide_str;
+
+        try {
+            _Wide_str = _STD wstring(static_cast<size_t>(_Num_chars_required), L'\0');
+        } catch (const _STD length_error&) {
+            return _STD unexpected{__std_win_error::_Insufficient_buffer};
+        } catch (...) {
+            return _STD unexpected{__std_win_error::_Not_enough_memory};
+        }
+
+        const int32_t _Num_chars_written = MultiByteToWideChar(
+            CP_UTF8, 0, _Str, static_cast<int>(_Str_size), _Wide_str.data(), static_cast<int>(_Num_chars_required));

```

Comment:
We are in `C++23`, so better use the `basic_string::resize_and_overwrite` for the best performance.